File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    # BFS 최솟값, 못가면 -1
    dx = [0,1,0,-1]
    dy = [1,0,-1,0]
    N = len(maps)
    M = len(maps[0])
    visited= [[False for _ in range(M)] for _ in range(N)]
    
    def BFS(i,j,n,m):
        q = deque()
        q.append((i,j,1))
        visited[i][j] = True

        while q:
            x, y, cnt = q.popleft()
            if x == n-1 and y == m-1:
                return cnt

            for k in range(4):
                nx, ny = x + dx[k], y + dy[k]
                if 0<=nx<n and 0<=ny<m and not visited[nx][ny] and maps[nx][ny] == 1:
                    visited[nx][ny] = True
                    q.append((nx,ny,cnt+1))
        return -1
    
    answer = BFS(0,0,N,M)
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = -1;
    const dir = [[0,1],[0,-1],[1,0],[-1,0]]
    const row = maps.length
    const col = maps[0].length
    const visited = Array.from({length:row}, ()=> Array.from({length:col}, () => false))
    
    const queue = []
    let begin = 0
    queue.push([0, 0, 1])
    while(begin < queue.length){
        const [cy, cx, cnt] = queue[begin++]
        if(cy === row - 1 && cx === col - 1) return cnt
        
        for(let k = 0; k < 4; k++){
            const ny = cy + dir[k][0]
            const nx = cx + dir[k][1]
            
            if(ny < 0 || ny >= row || nx < 0 || nx >= col || maps[ny][nx] === 0 || visited[ny][nx]) continue
            visited[ny][nx] = true
            queue.push([ny, nx, cnt + 1])
            
        }
    }
    
    return answer;
}